### PR TITLE
feat(abigen): emit cargo build script directives

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

... and an `include_bytes!` when not in build scripts. Signals Cargo to rebuild if the source files change.

## PR Checklist

-   [ ] Added Tests
-   [ ] Added Documentation
-   [ ] Breaking changes

### CODE REVIEW

File: ethers-contract/ethers-contract-abigen/src/lib.rs [hunk 1]
```diff
@@ -95,44 +115,31 @@ impl Abigen {
...
-        let name = file_name
-            .to_str()
-            .ok_or_else(|| eyre::eyre!("File name contains invalid UTF-8"))?
-            .split('.') // ignore everything after the first `.`
-            .next()
-            .unwrap(); // file_name is not empty as asserted by .file_name() already
-        let contents = fs::read_to_string(&path).wrap_err("Could not read file")?;
-
-        Self::new(name, contents)
+        let path = path.as_ref().to_str().ok_or_else(|| eyre::eyre!("invalid path"))?;
```
Comment (mattsse) [inline ethers-contract/ethers-contract-abigen/src/lib.rs:139]: nit: error should include the path

File: ethers-contract/ethers-contract-abigen/src/lib.rs [hunk 2]
```diff
@@ -202,12 +209,33 @@ impl Abigen {
         self
     }
 
+    /// Specify whether to print [cargo build script directives][ref] if the source is a path. By
+    /// default, this is true only when executing inside of a build script.
+    ///
+    /// [ref]: https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script
```
Comment (mattsse) [inline ethers-contract/ethers-contract-abigen/src/lib.rs:215]: I need to adapt this href style

File: ethers-contract/ethers-contract-abigen/src/lib.rs [hunk 3]
```diff
@@ -368,6 +396,11 @@ impl ContractBindings {
     }
 }
 
+/// Returns whether the current executable is a cargo build script.
+fn in_build_script() -> bool {
+    std::env::var("CARGO_MAKEFLAGS").is_ok()
```
Comment (mattsse) [inline ethers-contract/ethers-contract-abigen/src/lib.rs:401]: TIL

**Review by mattsse — APPROVED** [review]
nice, smol nit re error message otherwise lgtm

File: ethers-contract/ethers-contract-abigen/src/lib.rs (hunk 3 above)
Comment (DaniPopes) [reply ethers-contract/ethers-contract-abigen/src/lib.rs:401]: honestly i just went over all of the env vars, and this one was the only one that is set only on build scripts, per docs. Not 100% sure

Comment (DaniPopes) [reply ethers-contract/ethers-contract-abigen/src/lib.rs]: `TARGET` should be fine as well.

**Pull request merged**